### PR TITLE
Fixes #32187 - Import/Export for the file and ansible types

### PR DESCRIPTION
**Pull request by parthaa**
This commit adds support for import/export of file type. With this
commit one should be able to
1) Export Version/Library  file repos
2) Import Version/Library  file repos

3) Export Library for ansible_collection
4) Import Library for ansible_collections

### CODE REVIEW

**Comment by parthaa** [comment]
To test this:

- Create and sync a file repo -> https://fixtures.pulpproject.org/file/
- Add it to a content-view and publish
- Use the `hammer content-export complete library --organization=<...>` or `hammer content-export complete version --id=<...>` to export
- Cat the generated metadata json file and notice that the file repo you added is present there
- Now in a different org create a repo with the same name, product and content view
- Use  `hammer content-import library --organization=<importing> --path=... --metadata-file=....`  OR  `hammer content-import version  --content-view-id=<..>`
- After the import examine that 3 files are present in both the library repo  + if you imported version that new cvv should also have the 3 files in it.

**Comment by jlsherrill** [comment]
Partha asked me to look at the code an 👍🏻  from that perspective

File: app/models/katello/repository.rb [hunk 1]
```diff
@@ -134,7 +136,11 @@ class Repository < Katello::Model
     scope :order_by_root, ->(attr) { joins(:root).order("#{Katello::RootRepository.table_name}.#{attr}") }
     scope :with_content, ->(content) { joins(Katello::RepositoryTypeManager.find_content_type(content).model_class.repository_association_class.name.demodulize.underscore.pluralize.to_sym).distinct }
     scope :by_rpm_count, -> { left_joins(:repository_rpms).group(:id).order("count(katello_repository_rpms.id) ASC") } # smallest count first
-
+    scope :exportable, -> { with_type(EXPORTABLE_TYPES) }
+    scope :immediate_or_none, -> do
+      immediate.or(where("#{RootRepository.table_name}.download_policy" => nil)).
+        or(where("#{RootRepository.table_name}.download_policy" => ""))
```
Comment (jeremylenz) [inline app/models/katello/repository.rb:142]: Shouldn't we have a validation that makes it so we don't have to check both `nil` and `""`?

File: app/services/katello/pulp3/repository.rb [hunk 2]
```diff
@@ -282,6 +282,24 @@ def copy_units_by_href(unit_hrefs)
         tasks
       end
 
+      def copy_all(source_repository, mirror: false)
```
Comment (jeremylenz) [inline app/services/katello/pulp3/repository.rb:285]:
```suggestion
      def copy_all(source_repository:, mirror: false)
```

File: app/lib/actions/katello/content_view/publish.rb [hunk 3]
```diff
@@ -181,11 +184,15 @@ def version_for_publish(content_view, options)
         end
 
         def handle_import(version, path:, metadata:)
-          plan_action(::Actions::Pulp3::Orchestration::ContentViewVersion::Import, version, path: path, metadata: metadata)
-          plan_action(::Actions::Pulp3::Orchestration::ContentViewVersion::CopyVersionUnitsToLibrary, version)
-          concurrence do
-            version.importable_repositories.pluck(:id).each do |id|
-              plan_action(Katello::Repository::IndexContent, id: id)
+          return if @imported
+          @imported = true
```
Comment (jeremylenz) [inline app/lib/actions/katello/content_view/publish.rb:188]: Why was this needed?

I heard once that I shouldn't use class instance variables in dynflow actions, because you can't be 100% sure that it's always the same instance.  Instead it's more reliable to use `input` or `output`.  (Could probably do a `plan_self` at the end of `plan` and then check here for `input[:imported]`, something like that.

**Review by jeremylenz — COMMENTED** [review]
Thanks @parthaa 

Code looks good overall, a couple initial comments below.

Will try to finally get the testing started tomorrow :see_no_evil:

File: app/models/katello/repository.rb (hunk 1 above)
Comment (parthaa) [reply app/models/katello/repository.rb:142]: well file types have `""` and ansible I think has nil for download policy. I could add a rule that sets that column to nil if it  is ''. But would then have to create a migration for that. I didnt think this merited a migration but I am not averse to it.

**Comment by parthaa** [comment]
[test katello]

File: app/lib/actions/katello/content_view/publish.rb (hunk 3 above)
Comment (jeremylenz) [reply app/lib/actions/katello/content_view/publish.rb:188]:
```rb
handle_import(version, options.slice(:path, :metadata))
```
Looks like this doesn't even depend on `repositories`.  Can it be moved outside the `each` loop?

If not, I still think `input[:imported]` instead of `@imported` would work fine.  Once you've set an input within `plan`, you can query it as long as it doesn't depend on any other step's output.

As an alternative you could set a local variable (`imported = false`) outside the `each` loop and then set `imported = true` after.  Like you have above except the initialization is outside the loop.  That doesn't seem bad to me.

**Review by jeremylenz — COMMENTED** [review]
:heavy_check_mark: File import / export works as expected!

 ```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-export complete library --organization "Default Organization"
[..............................................................................................................................................................] [100%]
Unable to access/write to '/var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-09T17-21-36-00-00'. Generated '/home/vagrant/hammer-cli-katello/metadata-1.json' instead. This file is necessary to perform an import.

[vagrant@centos7-katello-devel hammer-cli-katello]$ sudo su
[root@centos7-katello-devel hammer-cli-katello]# ll /var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-09T17-21-36-00-00
total 12
-rw-r--r--. 1 pulp pulp 5984 Apr  9 17:21 export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721.tar.gz
-rw-r--r--. 1 pulp pulp  333 Apr  9 17:21 export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721-toc.json
[root@centos7-katello-devel hammer-cli-katello]# exit
exit

[vagrant@centos7-katello-devel hammer-cli-katello]$ cat /home/vagrant/hammer-cli-katello/metadata-1.json | jq
{
  "organization": "Default Organization",
  "repository_mapping": {
    "test1-18555": {
      "repository": "test1",
      "product": "Accountant Stuff",
      "redhat": false
    },
    "test2-22134": {
      "repository": "test2",
      "product": "Accountant Stuff",
      "redhat": false
    },
    "Accountant_Stuff-3530": {
      "repository": "Accountant Stuff",
      "product": "Accountant Stuff",
      "redhat": false
    }
  },
  "content_view": "Export-Library",
  "content_view_version": {
    "major": 1,
    "minor": 0
  },
  "incremental": false,
  "toc": "export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721-toc.json"
}
[vagrant@centos7-katello-devel hammer-cli-katello]$ sudo su
[root@centos7-katello-devel hammer-cli-katello]# ll /var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-09T17-21-36-00-00
total 12
-rw-r--r--. 1 pulp pulp 5984 Apr  9 17:21 export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721.tar.gz
-rw-r--r--. 1 pulp pulp  333 Apr  9 17:21 export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721-toc.json
[root@centos7-katello-devel hammer-cli-katello]# cp -r /var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-09T17-21-36-00-00 /var/lib/pulp/imports/jer_import
[root@centos7-katello-devel hammer-cli-katello]# ll /var/lib/pulp/imports
total 0
drwxr-xr-x. 2 root root 152 Apr  9 17:27 jer_import
[root@centos7-katello-devel hammer-cli-katello]# ll /var/lib/pulp/imports/jer_import/
total 12
-rw-r--r--. 1 root root 5984 Apr  9 17:27 export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721.tar.gz
-rw-r--r--. 1 root root  333 Apr  9 17:27 export-aa808df3-d7e6-4cf9-92e9-66ee77a7c2f7-20210409_1721-toc.json

[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization="Import Org" --path=/var/lib/pulp/imports/jer_import
Could not import the archive.:
  Error: Unable to find '/var/lib/pulp/imports/jer_import/metadata.json'. If the metadata.json file is at a different location provide it to the --metadata-file option 
  
  See: 'hammer content-import library --help'.
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization="Import Org" --path=/var/lib/pulp/imports/jer_import --metadata-file /home/vagrant/hammer-cli-katello/metadata-1.json
Could not import the archive.:
  The following repositories provided in the import metadata are either not available in the Library or are of incorrect Respository Type. Please add or enable the repositories before importing
   
  * Product = 'Accountant Stuff', Repository = 'test1', Repository Type = 'Custom'
  * Product = 'Accountant Stuff', Repository = 'test2', Repository Type = 'Custom'
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization="Import Org" --path=/var/lib/pulp/imports/jer_import --metadata-file /home/vagrant/hammer-cli-katello/metadata-1.json
[..............................................................................................................................................................] [100%]
```
file_repo

File: app/lib/actions/katello/content_view/publish.rb (hunk 3 above)
Comment (parthaa) [reply app/lib/actions/katello/content_view/publish.rb]: Updated :)

**Comment by jeremylenz** [comment]
Ansible collection export & import works as expected!

donut_collection

**Comment by jeremylenz** [comment]
^ However, my user experience was less than ideal.  Some more general comments below (don't have to be addressed here, but should think about them for import/export overall)

```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-export library --organization "Default Organization"
Error: No such sub-command 'library'.

See: 'hammer content-export --help'.
```
How much effort would it be to add an alias `library` > `complete library` ?

```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-export complete library --organization "Default Organization"
[...........................................................................................................................................................................] [100%]
Unable to access/write to '/var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-14T13-57-57-00-00'. Generated '/home/vagrant/hammer-cli-katello/metadata-1.json' instead. This file is necessary to perform an import.
[vagrant@centos7-katello-devel hammer-cli-katello]$ cat /home/vagrant/hammer-cli-katello/metadata-1.json | jq
{
  "organization": "Default Organization",
  "repository_mapping": {},
  "content_view": "Export-Library",
  "content_view_version": {
    "major": 1,
    "minor": 0
  },
  "incremental": false,
  "toc": "export-56150409-da39-4703-901d-d93b0862162c-20210414_1357-toc.json"
}
[vagrant@centos7-katello-devel hammer-cli-katello]$ sudo su -
Last login: Tue Apr 13 16:31:25 UTC 2021 on pts/1
[root@centos7-katello-devel ~]# ll /var/lib/pulp/imports
total 0
drwxr-xr-x. 2 root root 152 Apr  9 17:27 jer_import
drwxr-xr-x. 3 root root  39 Apr 12 19:46 jer_import_2
[root@centos7-katello-devel ~]# mkdir -p /var/lib/pulp/imports/jer_import_3
[root@centos7-katello-devel ~]# cp -r /var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-14T13-57-57-00-00 /var/lib/pulp/imports/jer_import_3
[root@centos7-katello-devel ~]# exit
logout
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-1.json
Could not import the archive.:
  Error: Option --path is required.
  
  See: 'hammer content-import library --help'.
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-1.json --path /var/lib/pulp/exports/Default_Organization/Export-Library/1.0/2021-04-14T13-57-57-00-00
Could not import the archive.:
  Missing values for repository_mapping.
```
Here I was giving it the wrong path (I mistakenly gave it the export path instead of the import path.) But the error just says `Missing values for repository_mapping`.  Can we add a prompt to the error like 
`Missing values for repository_mapping. Did you specify the correct path?`


```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-1.json --path /var/lib/pulp/imports/jer_import_3
Could not import the archive.:
  Missing values for repository_mapping.
[vagrant@centos7-katello-devel hammer-cli-katello]$ ll /var/lib/pulp/imports/jer_import_3
ls: cannot access /var/lib/pulp/imports/jer_import_3: Permission denied
[vagrant@centos7-katello-devel hammer-cli-katello]$ sudo !!
sudo ll /var/lib/pulp/imports/jer_import_3
sudo: ll: command not found
[vagrant@centos7-katello-devel hammer-cli-katello]$ sudo su
[root@centos7-katello-devel hammer-cli-katello]# ll /var/lib/pulp/imports/jer_import_3
total 0
drwxr-xr-x. 2 root root 152 Apr 14 13:59 2021-04-14T13-57-57-00-00
[root@centos7-katello-devel hammer-cli-katello]# exit
exit
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-1.json --path /var/lib/pulp/imports/jer_import_3/2021-04-14T13-57-57-00-00
Could not import the archive.:
  Missing values for repository_mapping.
```
this part was my fault, I had forgotten to switch branches :yum: 

```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-export complete library --organization "Default Organization"
[...........................................................................................................................................................................] [100%]
Unable to access/write to '/var/lib/pulp/exports/Default_Organization/Export-Library/2.0/2021-04-14T14-04-04-00-00'. Generated '/home/vagrant/hammer-cli-katello/metadata-2.json' instead. This file is necessary to perform an import.
[vagrant@centos7-katello-devel hammer-cli-katello]$ sudo su
[root@centos7-katello-devel hammer-cli-katello]# mkdir -p /var/lib/pulp/imports/jer_import_4
[root@centos7-katello-devel hammer-cli-katello]# cp -r /var/lib/pulp/exports/Default_Organization/Export-Library/2.0/2021-04-14T14-04-04-00-00 /var/lib/pulp/imports/jer_import_4
[root@centos7-katello-devel hammer-cli-katello]# ll /var/lib/pulp/imports/jer_import_4
total 0
drwxr-xr-x. 2 root root 152 Apr 14 14:04 2021-04-14T14-04-04-00-00
[root@centos7-katello-devel hammer-cli-katello]# exit
exit
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/
[..................................                                                                                                                                          ] [20%]
Error: [Errno 2] No such file or directory: '/var/lib/pulp/imports/jer_import_4/export-c77ae679-67f1-49d5-b401-7c4b9c77b972-20210414_1404-toc.json'
```
This simple human error (I forgot to add `/2021-04-14T14-04-04-00-00` to the end of the path) had a severe penalty:

```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/2021-04-14T14-04-04-00-00
Could not import the archive.:
  Required lock is already taken by other running tasks.
  Please inspect their state, fix their errors and resume them.
  
  Conflicts with tasks:
  - http://centos7-katello-devel.redhatlaptop.example.com/foreman_tasks/tasks/8be22b45-0cb2-43da-a1ba-f52a9cd12ae2
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/2021-04-14T14-04-04-00-00
Could not import the archive.:
  Required lock is already taken by other running tasks.
  Please inspect their state, fix their errors and resume them.
  
  Conflicts with tasks:
  - http://centos7-katello-devel.redhatlaptop.example.com/foreman_tasks/tasks/8be22b45-0cb2-43da-a1ba-f52a9cd12ae2
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/2021-04-14T14-04-04-00-00
Could not import the archive.:
  Required lock is already taken by other running tasks.
  Please inspect their state, fix their errors and resume them.
  
  Conflicts with tasks:
  - http://centos7-katello-devel.redhatlaptop.example.com/foreman_tasks/tasks/8be22b45-0cb2-43da-a1ba-f52a9cd12ae2
```
Took me several tries to properly clean up the task.  We _have_ to find a way to make this easier.

```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/2021-04-14T14-04-04-00-00
Could not import the archive.:

  Content View Version specified in the metadata - 'Import-Library 2.0' already exists. If you wish to replace the existing version, delete Import-Library 2.0 and try again. 
```
Manually deleting the version is easier than the task cleanup, but still a pain.  How much effort would it be to add a `--force-replace-version` option or something, so that it could do it for you?


```
[vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/2021-04-14T14-04-04-00-00
[...........................................................................................................................................................................] [100%]
[vagrant@centos7-katello-devel hammer-cli-katello]$ 
```

and finally, done!

**Comment by parthaa** [comment]
> ^ However, my user experience was less than ideal. Some more general comments below (don't have to be addressed here, but should think about them for import/export overall)
> ....
> [vagrant@centos7-katello-devel hammer-cli-katello]$ hammer content-import library --organization "Import Org" --metadata-file /home/vagrant/hammer-cli-katello/metadata-2.json --path /var/lib/pulp/imports/jer_import_4/2021-04-14T14-04-04-00-00
> [...........................................................................................................................................................................] [100%]
> [vagrant@centos7-katello-devel hammer-cli-katello]$ 
> ```
> 
> and finally, done!

Thank you for the comments. You raised important points here.  
The main reason you are having to clean up the tasks is because we are creating content view versions when we should not be. This should be addressed as a a part of https://github.com/Katello/katello/pull/9248


> How much effort would it be to add an alias `library` > `complete library` ?
I like this idea and hit that confusion plenty of times. I would like wait for feedback from Upstream Users/QE/Docs before thinking more about this. I don't like creating extra sub commands that aliases which might potentially confuse the user at this point.

**Review by jeremylenz — APPROVED** [review]
Tested both file and ansible types with `content-export complete library`
Not tested: `version` or `incremental`

with that said.. LGTM :+1:
